modules/build/src/main/scala/scala/build/interactive/Interactive.scala

```diff
@@ -1,44 +1,79 @@
 import scala.io.StdIn.readLine
 
 case class Interactive(options: BuildOptions) {
 
   private val isInteractive = options.internal.interactive.getOrElse(false)
 
-  def chooseOne[E <: BuildException](
-    msg: String,
-    options: List[String],
-    fallbackError: E
-  ): Either[E, String] =
-    if (isInteractive && coursier.paths.Util.useAnsiOutput()) {
+  sealed abstract class Action[V] extends Product with Serializable {
+    def msg: String
+    def action: Option[V]
+    final def run: Option[V] =
+      if (isInteractive && coursier.paths.Util.useAnsiOutput())
+        action
+      else None
+  }
+
+  case class ConfirmOperation(msg: String) extends Action[Boolean] {
+    override def action: Option[Boolean] = {
+      System.err.println(s"$msg [Y/n]")
+      val response = readLine()
+      if (response.toLowerCase == "y")
+        Some(true)
+      else {
+        System.err.println("Abort")
+        None
+      }
+    }
+  }
+
+  case class ChooseOne(msg: String, options: List[String])
+      extends Action[String] {
+    override def action: Option[String] = {
       System.err.println(msg)
       options.zipWithIndex.foreach {
         case (option, index) => System.err.println(s"[$index] $option")
       }
       val response      = readLine()
-      val inputIndexOpt = parseIndexInput(response, options.length)
-      inputIndexOpt.map(options(_)).toRight(fallbackError)
+      val inputIndexOpt = parseIndexInput(response, options.length - 1)
+      inputIndexOpt.map(options(_))
     }
-    else Left(fallbackError)
-
-  private def parseIndexInput(input: String, range: Int): Option[Int] = {
-    val indexOpt = input.toIntOption
-    indexOpt match {
-      case Some(index) =>
-        val isInRange = index <= range && index >= 0
-        if (isInRange) Some(index)
-        else {
-          System.err.println(
-            s"The input index number is invalid, integer value from 0 to $range is expected."
-          )
+
+    private def parseIndexInput(input: String, range: Int): Option[Int] = {
+      val indexOpt = input.toIntOption
+      indexOpt match {
+        case Some(index) =>
+          val isInRange = index <= range && index >= 0
+          if (isInRange) Some(index)
+          else {
+            System.err.println(
+              s"The input index number is invalid, integer value from 0 to $range is expected."
+            )
+            None
+          }
+        case _ =>
+          System.err.println(s"Unable to parse input: integer value from 0 to $range is expected.")
           None
-        }
-      case _ =>
-        System.err.println(s"Unable to parse input: integer value from 0 to $range is expected.")
-        None
+      }
     }
   }
+
+  def confirmOperation[E](
+    msg: String,
+    fallbackError: () => E
+  ): Either[E, Boolean] = ConfirmOperation(msg).run.toRight(fallbackError())
+
+  def chooseOne[E](
+    msg: String,
+    options: List[String],
+    fallbackError: E
```

Comment:
Maybe this could be by-name, just like `confirmOperation` above. That way, this could replace the override just below: users would just have to call `.merge` on the resulting `Either[Nothing, String]` to recover a `String`.

Reply:
Actually, it seems we don't even need to pass the fallback error: these methods can return `Option[…]`, and callers would then call `.getOrElse { … }` on it.